beacon_node/execution_layer/src/engine_api.rs

```diff
@@ -563,6 +594,110 @@ pub struct BlobsBundleV1<E: EthSpec> {
+            Self::Capella(payload) => payload.execution_payload.parent_hash,
+            Self::Deneb(payload) => payload.execution_payload.parent_hash,
+        }
+    }
+
+    pub fn block_hash(&self) -> ExecutionBlockHash {
+        match self {
+            Self::Merge(payload) => payload.execution_payload.block_hash,
+            Self::Capella(payload) => payload.execution_payload.block_hash,
+            Self::Deneb(payload) => payload.execution_payload.block_hash,
+        }
+    }
+
+    pub fn block_number(&self) -> u64 {
+        match self {
+            Self::Merge(payload) => payload.execution_payload.block_number,
+            Self::Capella(payload) => payload.execution_payload.block_number,
+            Self::Deneb(payload) => payload.execution_payload.block_number,
+        }
+    }
+
+    pub fn into_execution_payload(self) -> ExecutionPayload<E> {
+        map_new_payload_request_into_execution_payload!(self, |request, cons| {
+            cons(request.execution_payload)
+        })
+    }
+}
+
+impl<'a, E: EthSpec> TryFrom<BeaconBlockRef<'a, E>> for NewPayloadRequest<E> {
+    type Error = BeaconStateError;
+
+    fn try_from(block: BeaconBlockRef<'a, E>) -> Result<Self, Self::Error> {
+        match block {
+            BeaconBlockRef::Base(_) | BeaconBlockRef::Altair(_) => {
+                Err(Self::Error::IncorrectStateVariant)
+            }
+            BeaconBlockRef::Merge(block_ref) => Ok(Self::Merge(NewPayloadRequestMerge {
+                execution_payload: block_ref.body.execution_payload.execution_payload.clone(),
+            })),
+            BeaconBlockRef::Capella(block_ref) => Ok(Self::Capella(NewPayloadRequestCapella {
+                execution_payload: block_ref.body.execution_payload.execution_payload.clone(),
+            })),
+            BeaconBlockRef::Deneb(block_ref) => Ok(Self::Deneb(NewPayloadRequestDeneb {
+                execution_payload: block_ref.body.execution_payload.execution_payload.clone(),
+                versioned_hashes: block_ref
+                    .body
+                    .blob_kzg_commitments
+                    .iter()
+                    .map(kzg_commitment_to_versioned_hash)
+                    .collect(),
+                parent_beacon_block_root: block_ref.parent_root,
+            })),
+        }
+    }
+}
+
+impl<E: EthSpec> TryFrom<ExecutionPayload<E>> for NewPayloadRequest<E> {
+    type Error = BeaconStateError;
+
+    fn try_from(payload: ExecutionPayload<E>) -> Result<Self, Self::Error> {
+        match payload {
+            ExecutionPayload::Merge(payload) => Ok(Self::Merge(NewPayloadRequestMerge {
+                execution_payload: payload,
+            })),
+            ExecutionPayload::Capella(payload) => Ok(Self::Capella(NewPayloadRequestCapella {
+                execution_payload: payload,
+            })),
+            ExecutionPayload::Deneb(_) => Err(Self::Error::IncorrectStateVariant),
```

Comment:
Is there a reason why this returns an error?

Reply:
For `deneb`, you can't convert only an `ExecutionPayload` into a `NewPayloadRequest`, you also need the `parent_beacon_block_root`. This conversion is only used for tests thus far. Lighthouse is always using the code above to convert from a `BeaconBlockRef` to a `NewPayloadRequest`.

Reply:
Related to this: https://github.com/ethereum/consensus-specs/pull/3454 
it's a pain point for lodestar I guess, has it impacted our tests?

Reply:
This is just here to make some testing code (which currently doesn't bother to test post-merge forks) more concise. It only impacts some merge fork tests (thus the error condition isn't triggered). There's `TODO`'s in the code to consider making those tests fork-aware / test multiple forks. I we decide to do that, this code will be revisited.

Reply:
If it's only used in tests, I wonder if it's worth adding `#[cfg(test)]` to make it more explicit and ensure we don't later use this in production code?

Reply:
Unfortunately we don't have our code structured that way. It's used in the `mock_execution_engine` as well as the execution engine integration tests. I believe the later isn't run with `cfg(test)` enabled. Fixing up these tests is beyond the scope of this PR. I'd be happy to look into doing this but I'm not sure it should be done before we get `deneb-free-blobs` merged into `unstable` as it would only make the review even worse.

Reply:
As it stands it's not really the worst thing in the world. I would be worried if it were a panic but an error isn't bad.

Reply:
ah yeah that wouldn't work for integration tests, got it thanks :)